### PR TITLE
feat: add dynamic config decoder

### PR DESCRIPTION
**Pull request by andrewrynhard**
This adds the ability to dynamically decode mult-doc YAML files.

Signed-off-by: Andrew Rynhard <andrew@rynhard.io>

---
This change is [Reviewable](https://reviewable.io/reviews/talos-systems/talos/2376)

### CODE REVIEW

File: pkg/config/configloader/configloader.go [hunk 1]
```diff
@@ -9,76 +9,50 @@ import (
...
-// NewFromBytes will take a byteslice and attempt to parse a config file from it.
-func NewFromBytes(in []byte) (config.Provider, error) {
-	c, err := fromBytes(in)
-	if err != nil {
-		return nil, err
+	// Look for the older flat v1alpha1 file first, since we have to handle it in
+	// a special way.
+	for _, manifest := range manifests {
+		if talosconfig, ok := manifest.(*v1alpha1.Config); ok {
+			log.L.Println("WARNING: This version of the machine config will be deprecated in a future release. Please prepare to migrate to the next version of the config as soon as possible.")
```
Comment (smira) [inline pkg/config/configloader/configloader.go]: should we not include this warning yet for 0.6?

File: pkg/config/decoder/decoder.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+	var (
+		version string
+		kind    string
+		spec    *yaml.Node
+	)
+
+	for i, node := range manifest.Content {
+		switch node.Value {
+		case ManifestVersionKey:
+			if err = manifest.Content[i+1].Decode(&version); err != nil {
```
Comment (smira) [inline pkg/config/decoder/decoder.go:107]: should we check that `i+1` is available? not sure, but what if document looks like `spec:` without any content on the right?

Comment (andrewrynhard) [reply pkg/config/decoder/decoder.go:107]: Probably a good test case. I will add it.

**Pull request merged**